File: src/main/java/com/oycm/algorithm/e/ea/basic/Solution_15.scala
package com.oycm.algorithm.e.ea.basic

object Solution_15 {

  /**
   * 3371. 识别数组中的最大异常值 1644
   * https://leetcode.cn/problems/identify-the-largest-outlier-in-an-array/description/
   *
   * 恰好 有 n - 2 个元素是 特殊数字
   *  - 一个元素是 所有 特殊元素的 和
   *  - 另一个是异常值
   *
   * @param nums 整数数组, n [3, 100 000], nums(i) [-1000, 1000]
   * @return 求 nums 中最大的异常值
   */
  def getLargestOutlier(nums: Array[Int]): Int = {
    /*
    数组和 = 特殊元素 * 2 + 异常值

    数组和 - nums(i) * 2 的这个值是否在数组中存在，存在则是异常值
    是否需要对结果进行判断？
    如果 exception == sum
    sums = 3 sum 如果这个值存在，则元素中至少有两个相同的值 30 10 235 235
    所以要用 map 判断次数
     */
    var ans = -1000
    var sums = 0
    val map = scala.collection.mutable.Map[Int,Int]()
    for (num <- nums) {
      sums += num
      map(num) = map.getOrElse(num, 0) + 1
    }
    // 枚举 元素和，找异常值
    for (num <- nums) {
      val sum = sums - 2 * num
      if (map.contains(sum) && (sum != num || map(sum) > 1)) {
        ans = Math.max(ans, sums - 2 * num)
      }
    }

    ans
  }

  def main(args: Array[String]): Unit = {
//    println(getLargestOutlier(Array(-2, -1, -3, -6, 4)))
//    println(getLargestOutlier(Array(1, 1, 1, 1, 1, 5, 5)))
//    println(getLargestOutlier(Array(2, 3, 5, 10)))
    println(getLargestOutlier(Array(6, -31, 50, -35, 41, 37, -42, 13)))
  }

}

File: src/main/java/com/oycm/month2026/august/Solution_15.java
package com.oycm.month2026.august;

public class Solution_15 {

    /**
     * 3702. 按位异或非零的最长子序列
     *
     * @param nums
     * @return
     */
    public int longestSubsequence(int[] nums) {
        /*
        给你一个整数数组 nums。
        返回 nums 中 按位异或（XOR）计算结果 非零 的 最长子序列 的长度。如果不存在这样的 子序列 ，返回 0 。
        子序列 是一个 非空 数组，可以通过从原数组中删除一些或不删除任何元素（不改变剩余元素的顺序）派生而来。
         */
        /*
        计算整个数组的异或和，如果不为 0，答案为 n；
        如果异或和为 0，去掉一个非 0 元素，就能得到最长子序列；
         */
        int mx = Integer.MIN_VALUE;
        int xor = 0;
        for (int x : nums) {
            xor ^= x;
            mx = Math.max(mx, x);
        }
        if (xor > 0) return nums.length;
        return mx == 0 ? 0 : nums.length - 1;
    }

}
